examples/Generic/README.md

```diff
@@ -2,21 +2,160 @@
 
 This directory contains a minimal example that shows how to use `AstarteGenericDevice`.
 
-The device is instantiated in `ExampleDevice.java`, the code is commented to highlight all the
-things needeed to create your Astarte Device.
+The device is instantiated in `ExampleDevice`, the code is commented to highlight all the
+things needed to create your Astarte Device.
+
+
+## Downloading
+
+Add to "dependencies" (as for any module you need):
+
+``` groovy
+dependencies {
+    implementation 'org.astarte-platform:devicesdk-generic:$version'
+}
+```
+
+Usage
+-----
+
+### 1. Device registration and credentials secret emission
+
+*If your device is already register, you can skip this section.*
+
+You must register the device beforehand to obtain its `credentials-secret`.
+You have three options to achieve this:
+- Using the astartectl command [`astartectl`](https://github.com/astarte-platform/astartectl).
+- Using the Astarte Dashboard, which is located at `https://dashboard.<your-astarte-domain>`.
+- Using `AstartePairingService` class contained in sdk (see following paragraph for details).
+
+#### 1.1 Programmatically Generate an Astarte Device ID
+
+You can generate an astarte device id either randomly or in a deterministic fashion, using a UUID namespace and some unique data.
+
+##### 1.1.1 Random
+
+``` java
+String deviceId = AstarteDeviceIdUtils.generateId();
+```
+
+##### 1.1.2 Deterministic (UUIDv5)
+
+``` java
+UUID uuidNamespace = UUID.fromString("f79ad91f-c638-4889-ae74-9d001a3b4cf8");
+String macAddress = "98:75:a8:0d:96:db";
+deviceId = AstarteDeviceIdUtils.generateId(uuidNamespace, macAddress);
+```
+
+#### 1.2. Register device
+
+Register a device, obtaining its credentials secret. To access the registration API, you must pass a valid JWT.
+``` java
+String credentialsSecret = new AstartePairingService(pairingUrl, realm).registerDevice(jwt, deviceId);
+```
+
+### 2. Astarte device creation
+
+We use h2 in memory as persistence in this example, but any DB compatible
+with JDBC can be used.
+
+Check `ExampleInterfaceProvider` to see which interfaces are supported by the device 
+and how to add more interfaces to your examples.
+
+``` java
+JdbcConnectionSource connectionSource = new JdbcConnectionSource("jdbc:h2:mem:testDb");
+AstarteDevice device =
+    new AstarteGenericDevice(
+        deviceId,
+        realm,
+        credentialsSecret,
+        new ExampleInterfaceProvider(),
+        pairingUrl,
+        connectionSource);
+```
+
+### 3. Connection listeners
+
+Check `ExampleMessageListener` to see how to listen for device connection, disconnection and failure.
```

Comment:
I would write something like
See `ExampleMessageListener` for further details about listening for device connection, disconnection and failure.